Mathlib/RingTheory/Coalgebra/Hom.lean

```diff
@@ -0,0 +1,310 @@
+  LinearMap.toAddMonoidHom_injective.comp coe_linearMap_injective
+
+protected theorem congr_fun {φ₁ φ₂ : A →ₗc[R] B} (H : φ₁ = φ₂) (x : A) : φ₁ x = φ₂ x :=
+  DFunLike.congr_fun H x
+
+protected theorem congr_arg (φ : A →ₗc[R] B) {x y : A} (h : x = y) : φ x = φ y :=
+  DFunLike.congr_arg φ h
+
+@[ext]
+theorem ext {φ₁ φ₂ : A →ₗc[R] B} (H : ∀ x, φ₁ x = φ₂ x) : φ₁ = φ₂ :=
+  DFunLike.ext _ _ H
+
+theorem ext_iff {φ₁ φ₂ : A →ₗc[R] B} : φ₁ = φ₂ ↔ ∀ x, φ₁ x = φ₂ x :=
+  DFunLike.ext_iff
+
+@[ext high]
+theorem ext_of_ring {f g : R →ₗc[R] A} (h : f 1 = g 1) : f = g :=
+  coe_linearMap_injective (by ext; assumption)
+
+@[simp]
+theorem mk_coe {f : A →ₗc[R] B} (h₁ h₂ h₃ h₄) : (⟨⟨⟨f, h₁⟩, h₂⟩, h₃, h₄⟩ : A →ₗc[R] B) = f :=
+  ext fun _ => rfl
+
+variable (R A)
+
+/-- Identity map as a `CoalgHom`. -/
+@[simps!] protected def id : A →ₗc[R] A :=
+  { LinearMap.id with
+    counit_comp := by ext; rfl
+    map_comp_comul := by simp only [map_id, LinearMap.id_comp, LinearMap.comp_id] }
+
+variable {R A}
+
+@[simp]
+theorem coe_id : ⇑(CoalgHom.id R A) = id :=
+  rfl
+
+@[simp]
+theorem id_toLinearMap : (CoalgHom.id R A : A →ₗ[R] A) = LinearMap.id := rfl
+
+/-- Composition of coalgebra homomorphisms. -/
+@[simps!] def comp (φ₁ : B →ₗc[R] C) (φ₂ : A →ₗc[R] B) : A →ₗc[R] C :=
+  { (φ₁ : B →ₗ[R] C) ∘ₗ (φ₂ : A →ₗ[R] B) with
+    counit_comp := by ext; simp
+    map_comp_comul := by ext; simp [map_comp] }
+
+@[simp]
+theorem coe_comp (φ₁ : B →ₗc[R] C) (φ₂ : A →ₗc[R] B) : ⇑(φ₁.comp φ₂) = φ₁ ∘ φ₂ := rfl
+
+@[simp]
+theorem comp_toLinearMap (φ₁ : B →ₗc[R] C) (φ₂ : A →ₗc[R] B) :
+    φ₁.comp φ₂ = (φ₁ : B →ₗ[R] C) ∘ₗ (φ₂ : A →ₗ[R] B) := rfl
+
+variable (φ : A →ₗc[R] B)
+
+@[simp]
+theorem comp_id : φ.comp (CoalgHom.id R A) = φ :=
+  ext fun _x => rfl
+
+@[simp]
+theorem id_comp : (CoalgHom.id R B).comp φ = φ :=
+  ext fun _x => rfl
+
+theorem comp_assoc (φ₁ : C →ₗc[R] D) (φ₂ : B →ₗc[R] C) (φ₃ : A →ₗc[R] B) :
+    (φ₁.comp φ₂).comp φ₃ = φ₁.comp (φ₂.comp φ₃) :=
+  ext fun _x => rfl
+
+theorem map_smul_of_tower {R'} [SMul R' A] [SMul R' B] [LinearMap.CompatibleSMul A B R' R] (r : R')
+    (x : A) : φ (r • x) = r • φ x :=
+  φ.toLinearMap.map_smul_of_tower r x
+
+@[simps (config := .lemmasOnly) toSemigroup_toMul_mul toOne_one]
+instance End : Monoid (A →ₗc[R] A) where
+  mul := comp
+  mul_assoc ϕ ψ χ := rfl
+  one := CoalgHom.id R A
+  one_mul ϕ := ext fun x => rfl
+  mul_one ϕ := ext fun x => rfl
+
+@[simp]
+theorem one_apply (x : A) : (1 : A →ₗc[R] A) x = x :=
+  rfl
+
+@[simp]
+theorem mul_apply (φ ψ : A →ₗc[R] A) (x : A) : (φ * ψ) x = φ (ψ x) :=
+  rfl
+
+end
+
+section AddCommGroup
+
+variable [CommSemiring R] [AddCommGroup A] [AddCommGroup B] [Module R A] [Module R B]
+
+variable [CoalgebraStruct R A] [CoalgebraStruct R B] (φ : A →ₗc[R] B)
+
+protected theorem map_neg (x) : φ (-x) = -φ x :=
+  map_neg _ _
+
+protected theorem map_sub (x y) : φ (x - y) = φ x - φ y :=
+  map_sub _ _ _
```

Comment:
Do you need these?

Reply:
Nope. There are loads of similar `protected` theorems in `Algebra.Algebra.Hom` I chose not to emulate and I didn't intend to keep these ones in particular.